README.md
# Titlekit [![Build Status](https://travis-ci.org/simonrepp/titlekit.png)](https://travis-ci.org/simonrepp/titlekit)

Featureful Ruby library for SRT / ASS / SSA subtitles

*Titlekit supports SRT, ASS and SSA, file format conversion, transcoding, automatic encoding detection, creation of simultaneous/multi-language subtitles, and timecode corrections with simple, progressive and framerate-based approaches. All of this is packed into a natural, dead-simple (and also irb-friendly) API.*

## Installation

Add the `titlekit` gem to your gemfile or install it yourself:

     $ gem install titlekit
     
### Optional installation recommendation

Titlekit uses [rchardet19](https://github.com/oleander/rchardet) to detect unknown encodings; There is another library for this task called [charlock_holmes](https://github.com/brianmario/charlock_holmes), which offers more robust detection algorithms, but is not included by default because it relies on external C libraries that can make its installation anywhere from semi-easy to impossible. If you want Titlekit to use *charlock_holmes*, install it yourself, and Titlekit will automatically use it over *rchardet19*!

## Documentation

### Basic example

A small hello world of Titlekit: **Converting from .srt to .ssa format**

```ruby
  job = Titlekit::Job.new             # (1) Initialize
  job.have { file('existing.srt') }   # (2) Specify what you have
  job.want { file('converted.ssa') }  # (3) Specify what you want
  job.run                             # (4) Make it happen
```
### Checking success

The return value from `#run` will tell you if the job was a success. If it was not,
you can access `#report` to get messages related to the direct failure cause and also on
anything suspicious that might have happened before (e.g. Low confidence when detecting an
unknown encoding)

```ruby
  if job.run
    # hooray
  else
    puts job.report.join("\n")
  end
```

### Commandline Template Generator

To get you started quickly with edit scripts for subtitles, you can use a console command provided by Titlekit:

     $ titlekit templates

If there are no subtitle files present in your current directory, this will generate a file called `process_generic.rb`, if there are subtitle files present, it will generate a file called `process_[your-subtitle-filename].rb` for each subtitle file that was found. In either case all of these files will contain a basic but complete scaffolding script for processing your subtitle file with Titlekit. In the case of the non-generic templates, they will already contain the correct filename for the input file. Together with some concise in-line comments, these templates are meant to help you quickly start off with your task.

### All features by example

In all following examples I will omit `(1)` and `(4)` from the basic example, because they stay the same.  
All the functionalities from all the following examples can be combined in any way you want.

#### Transcoding

```ruby
  job.have do
    file('input.srt')
    encoding('ISO-8859-1')
  end
  
  job.want do
    file('output.srt')
    encoding('UTF-8')
  end
```

#### Converting

```ruby
  job.have { file('input.ass') }
  job.want { file('output.srt') }
```

#### Simple timeshifting

```ruby
  job.have do
    file('input.srt')
    reference('first sentence spoken', subtitle: 0)
  end
  
  job.want do
    file('output.srt')
    reference('first sentence spoken', srt_timecode: '00:00:54,200')
  end
```

#### Progressive timeshifting

```ruby
  job.have do
    file('input.srt')
    reference('first subtitle', subtitle: 0)
    reference('last subtitle', subtitle: 475)
  end
  
  job.want do
    file('output.srt')
    reference('first subtitle', minutes: 3.76)
    reference('last subtitle', hours: 1.8912)
  end
```

#### Framerate-based timeshifting

```ruby
  job.have do
    file('input.srt')
    fps(25)
  end
  
  job.want do
    file('output.srt')
    fps(23.976)
  end
```

#### Mixed mode timeshifting

```ruby
  job.have do
    file('input.srt')
    fps(25)
    reference(:first_sub, subtitle: 0)
  end
  
  job.want do
    file('output.srt')
    fps(23.976)
    reference(:first_sub, minute: 13.49)    
  end
```

#### Merging

Subtitles that don't overlap are automatically merged and treated as one track.

```ruby
  job.have do
    file('subs_chapter1.srt')
  end

  job.have do
    file('subs_chapter2.srt')
  end

  job.want do
    file('subs_both_chapters_combined.srt')
  end
```

#### Merging (with time correction)

If each of your individual subtitle files contains timecodes relative to its own starting point,
you have to supply a reference so misaligned subtitles can be automatically shifted. When you fail
to do this, your subtitles overlap and thus Titlekit assumes that you want simultaneous subtitles (which are explained in the next paragraph below this one).

```ruby
  job.have do
    file('subs_cd1.srt')
  end

  job.have do
    file('subs_cd2.srt')
    reference(:cd2_subtitles_starting_at, minutes: 0)
  end

  job.want do
    reference(:cd2_subtitles_starting_at, srt_timecode: '00:01:24,000')
    file('subs_both_cds_combined.srt')
  end
```

#### Simultaneous/multi-lanugage subtitles

Subtitles that overlap are automatically treated as simultaneous/multi-lanugage subtitles.
Titlekit then positions and formats them in the most sensible way it sees fit.

```ruby
  job.have { file('enlish.srt') }
  job.have { file('dutch.srt') }
  job.want { file('dual-language.srt') }
```

Any target format is possible, Titlekit will automatically make the best use of the formatting
features your target format provides. Pick a sophisticated subtitle format, and your dual
subtitles will automatically be prettier and more readable!

```ruby
  job.have { file('enlish.srt') }
  job.have { file('dutch.srt') }
  job.want { file('dual-language-prettier.ass') }
```

You can also go crazy if you want, Titlekit can handle it.

```ruby
  job.have { file('enlish.srt') }
  job.have { file('dutch.srt') }
  job.have { file('spanish.srt') }
  job.have { file('german.srt') }
  job.have { file('italian.srt') }
  job.have { file('french.srt') }
  job.have { file('portuguese.srt') }
  job.have { file('russian.srt') }
  job.want { file('messy-but-supported.ssa') }
```

#### Mixed mode Merging (implicit)

If you really need the absurdly exotic case of merging multi-part subtitles into multiple simultaneous tracks, you just need to make sure you enter them in the correct order, which is: Lanuage1/Part1 -> Language1/Part2 -> Language2/Part1 -> Language2/Part2

```ruby
  job.have { file('subs_english_chapter1.srt')
  job.have { file('subs_english_chapter2.srt')
  job.have { file('subs_french_chapter1.srt')
  job.have { file('subs_french_chapter2.srt')

  job.want do
    file('subs_multi_part_multi_track.srt')
  end
```

#### Mixed mode Merging (explicit)

If you supply explicit track identifiers by which to material should be grouped into tracks, 
you can also forget all about the otherwise required order and just go wild:

```ruby

  job.have do
    file('subs_french_chapter2.srt')
    track('le-french-track')
  end

  job.have do
    file('subs_english_chapter1.srt')
    track('the-english-one')
  end

  job.have do
    file('subs_french_chapter1.srt')
    track('le-french-track')
  end

  job.have do
    file('subs_english_chapter2.srt')
    track('the-english-one')
  end

  job.want do
    file('subs_multi_part_multi_track.srt')
  end
```

#### Multiple targets

```ruby
  job.have { file('input.srt') }
  job.want { file('output.ass') }
  job.want { file('output.ssa') }
```

#### Templates

```ruby
  job.have do
    file('input.srt')
    encoding('Shift_JIS')
    reference(:some_subtitle, subtitle: 23)
  end
  
  templ = job.want do
    file('output.srt')
    encoding('UTF-8')
    reference(:some_subtitle, hours: 0.16)
  end
  
  job.want(template: templ) { file('output.ass') }
  job.want(template: templ) { file('output.ssa') }
```

#### Explicitly control encoding detection

```ruby
  job.have do
    file('input.srt')
    encoding(:detect) # Detect the encoding with charlock_holmes if installed, otherwise rchardet19
                      # You don't need to supply this line though, it's the default behavior!
  end

  job.have do
    file('input.srt')
    encoding(:rchardet19) # Explicitly use rchardet19
  end  
  
  job.have do
    file('input.srt')
    encoding(:charlock_holmes) # Explicitly use charlock_holmes
  end
```

#### Syntax Variants

`#have` and `#want` offer three different syntax variants, which are functionally identical:

```ruby
  job.have do
    file('input.srt')
    encoding('ISO-8859-1')
  end
  
  # is identical to
  
  job.have do |have|
    have.file('input.srt')
    have.encoding('ISO-8859-1')
  end
  
  # is identical to
  
  have = job.have
  have.file('input.srt')
  have.encoding('ISO-8859-1')
```

## API Reference

http://www.rubydoc.info/gems/titlekit/frames (or generate it yourself with YARD)

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# Changelog

This project follows semantic versioning.

## 1.2.3

- Fix inequalities caused by float addition (William T. Nelson) `8c84718`

## <= 1.2.2

*See commit history*


lib/titlekit/job.rb
module Titlekit
  class AbortJob < StandardError
  end

  class Job

    # Returns everything you {Have}
    #
    # @return [Array<Have>] Everything you {Have}
    attr_reader :haves

    # Returns everything you {Want}
    #
    # @return [Array<Want>] Everything you {Want}
    attr_reader :wants

    # Returns the job report, which documents the direct cause of failures
    # and any other unusual events that occur on the job. (regardless if it
    # failed or succeeded)
    #
    # @return [Array<String>] All reported messages
    attr_reader :report

    # Starts a new job.
    #
    # A job requires at least one file you {Have} and one file you {Want}
    # in order to be runable. Use {Job#have} and {Job#want} to add
    # and obtain specification interfaces for the job.
    #
    def initialize
      @haves = []
      @wants = []
      @report = []

      require 'rchardet19'

      begin
        if Gem::Specification.find_by_name('charlock_holmes')
          require 'charlock_holmes'
        end
      rescue Gem::LoadError
      end
    end

    # Runs the job.
    #
    # @return [Boolean] true if the job succeeds, false if it fails.
    #   {Job#report} provides information in case of failure.
    def run
      @wants.each do |want|
        @haves.each do |have|
          import(have)
          retime(have, want)
          cull(have)
          group(have)

          want.subtitles += have.subtitles.clone
        end

        polish(want)
        export(want)
      end

      return true
    rescue AbortJob
      return false
    end

    # Adds a new {Have} specification to your job.
    #
    # @example Using a block without a variable (careful: the scope changes!)
    #   job.have do
    #     encoding('utf-8')
    #     file('path/to/my/input.srt')
    #     fps(25)
    #   end
    #
    # @example Using a block and providing a variable
    #   job.have do |have|
    #     have.encoding('utf-8')
    #     have.file('path/to/my/input.srt')
    #     have.fps(25)
    #   end
    #
    # @example Catching the reference and assigning things at any later point
    #   have = job.have
    #   have.encoding('utf-8')
    #   have.file('path/to/my/input.srt')
    #   have.fps(25)
    #
    # @example Cloning a previous specification and extending on it
    #   have2 = job.have(template: have1)
    #   have2.encoding('ISO-8859-1')
    #   have2.file('path/to/my/input2.srt')
    #
    # @param template [Have] optionally you can specify another {Have} as a
    #   template, from which all properties but the file path are cloned
    # @return [Have] a reference to the newly assigned {Have}
    def have(*_args, template: nil, &block)
      specification = Have.new

      if template
        specification.fps = template.fps.clone
        specification.references = template.references.clone
      end

      if block
        if block.arity < 1
          specification.instance_eval(&block)
        else
          block[specification]
        end
      end

      @haves << specification

      specification
    end

    # Adds a new {Want} specification to your job.
    #
    # @example Using a block without a variable (careful: the scope changes!)
    #   job.want do
    #     encoding('utf-8')
    #     file('path/to/my/output.srt')
    #     fps(23.976)
    #   end
    #
    # @example Using a block and providing a variable
    #   job.want do |want|
    #     want.encoding('utf-8')
    #     want.file('path/to/my/output.srt')
    #     want.fps((23.976)
    #   end
    #
    # @example Catching the reference and assigning things at any later point
    #   want = job.want
    #   want.encoding('utf-8')
    #   want.file('path/to/my/output.srt')
    #   want.fps((23.976)
    #
    # @example Cloning a previous specification and extending on it
    #   want2 = job.want(template: want1)
    #   want2.encoding('ISO-8859-1')
    #   want2.file('path/to/my/output.ass')
    #
    # @param template [Want] optionally you can specify another {Want} as a
    #   template, from which all properties but the file path are cloned
    # @return [Want] a reference to the newly assigned {Want}
    def want(*args, template: nil, &block)
      specification = Want.new

      if template
        specification.fps = template.fps.clone
        specification.references = template.references.clone
      end

      if block
        if block.arity < 1
          specification.instance_eval(&block)
        else
          block[specification]
        end
      end

      @wants << specification

      specification
    end

    private

    # Imports what we {Have}
    #
    # @param [Have] What we {Have}
    def import(have)
      begin
        data = File.read(have.file)
      rescue
        @report << "Failure while reading #{have.file}"
        raise AbortJob
      end

      begin
        if [:detect, :charlock_holmes].include?(have.encoding) && defined?(CharlockHolmes)
          detection = CharlockHolmes::EncodingDetector.detect(data)
          @report << "Assuming #{detection[:encoding]} for #{have.file} (detected by charlock_holmes with #{detection[:confidence]}% confidence)"
          data.force_encoding(detection[:encoding])
        elsif [:detect, :rchardet19].include?(have.encoding) && defined?(CharDet)
          detection = CharDet.detect(data)
          @report << "Assuming #{detection.encoding} for #{have.file} (detected by rchardet19 with #{(detection.confidence * 100).to_i}% confidence)"
          data.force_encoding(detection.encoding)
        else
          @report << "Assuming #{have.encoding} for #{have.file} (user-supplied)"
          data.force_encoding(have.encoding)
        end
      rescue
        @report << "Failure while setting encoding for #{have.file}"
        raise AbortJob
      end

      begin
        data.encode!('UTF-8')
      rescue
        @report << "Failure while transcoding #{have.file} from #{data.encoding} to intermediate UTF-8 encoding"
        raise AbortJob
      end

      begin
        have.subtitles =
          case File.extname(have.file)
          when '.ass'
            ASS.import(data)
          when '.ssa'
            SSA.import(data)
          when '.srt'
            SRT.import(data)
          when '.yt'
            YT.import(data)
          else
            fail 'Not supported'
          end
      rescue
        @report << "Failure while importing #{File.extname(have.file)[1..3].upcase} from #{have.file}"
        raise AbortJob
      end
    end

    # Transfers the subtitles from the state we {Have}, to the state we {Want}.
    #
    # @params have [Have] What we {Have}
    # @params want [Want] What we {Want}
    def retime(have, want)
      matching_references = want.references.keys & have.references.keys

      # Resolve subtitle references by getting actual timecodes
      matching_references.each do |reference|
        if (index = have.references[reference][:subtitle])
          have.references[reference][:timecode] = have.subtitles[index][:start]
        end
      end

      case matching_references.length
      when 3..(_infinity = 1.0 / 0)
        # "synchronization jitter" correction by interpolating ? Consider !
      when  2
        retime_by_double_reference(have,
                                   want,
                                   matching_references[0],
                                   matching_references[1])
      when 1
        if have.fps && want.fps
          retime_by_framerate_plus_reference(have, want, matching_references[0])
        else
          retime_by_single_reference(have, want, matching_references[0])
        end
      when 0
        retime_by_framerate(have, want) if have.fps && want.fps
      end
    end

    # Cleans out subtitles that fell out of the usable time range
    #
    # @params have [Have] What we {Have}
    def cull(have)
      have.subtitles.reject! { |subtitle| subtitle[:end] < 0 }
      have.subtitles.each do |subtitle|
        subtitle[:start] = 0 if subtitle[:start] < 0
      end
    end

    # Assigns track identification fields for distinguishing
    # between continuous/simultaneous subtitles
    #
    # @params have [Have] What we {Have}
    def group(have)
      if have.track
        # Assign a custom track identifier if one was supplied
        have.subtitles.each { |subtitle| subtitle[:track] = have.track }
      elsif @haves.index(have) == 0 || @haves[@haves.index(have) - 1].subtitles.empty?
        # Otherwise let the path be the track identifier for the first subtitle
        have.subtitles.each { |subtitle| subtitle[:track] = have.file }
      else
        # For the 2nd+ subtitles determine the track association by detecting
        # collisions against the previously imported subtitles

        collisions = 0

        have.subtitles.each do |subtitle|
          @haves[@haves.index(have) - 1].subtitles.each do |previous_subtitle|
            collisions += 1 if (subtitle[:start] > previous_subtitle[:start] &&
                                subtitle[:start] < previous_subtitle[:end]) ||
                               (subtitle[:end] > previous_subtitle[:start] &&
                                subtitle[:end] < previous_subtitle[:end]) ||
                               (previous_subtitle[:start] > subtitle[:start] &&
                                previous_subtitle[:start] < subtitle[:end]) ||
                               (previous_subtitle[:end] > subtitle[:start] &&
                                previous_subtitle[:end] < subtitle[:end]) ||
                               (subtitle[:start] == previous_subtitle[:start] ||
                                subtitle[:end] == previous_subtitle[:end])
          end
        end

        if collisions.to_f / have.subtitles.length.to_f > 0.01
          # Add a new track if there are > 1% collisions between these
          # subtitles and the ones that were last imported
          have.subtitles.each { |subtitle| subtitle[:track] = have.file }
        else
          # Otherwise continue using the previous track identifier
          # (= Assume that these and the previous subtitles are one track)
          previous_track = @haves[@haves.index(have) - 1].subtitles.first[:track]
          have.subtitles.each { |subtitle| subtitle[:track] = previous_track }
        end
      end
    end

    # Polishes what we {Want}
    #
    # @params want [Want] What we {Want} polished
    def polish(want)
      # Glue subtitle starts
      want.subtitles.sort_by! { |subtitle| subtitle[:start] }
      want.subtitles.each_cons(2) do |pair|
        distance = pair[1][:start] - pair[0][:start]
        if distance < want.glue_treshold
          pair[0][:start] += distance / 2
          pair[1][:start] -= distance / 2
        end
      end

      # Glue subtitles ends
      want.subtitles.sort_by! { |subtitle| subtitle[:end] }
      want.subtitles.each_cons(2) do |pair|
        if pair[1][:end] - pair[0][:end] < want.glue_treshold
          pair[0][:end] += (pair[1][:end] - pair[0][:end]) / 2
          pair[1][:end] -= (pair[1][:end] - pair[0][:end]) / 2
        end
      end
    end

    # Exports what we {Want}
    #
    # @param want [Want] What we {Want}
    def export(want)
      begin
        data =
          case File.extname(want.file)
          when '.ass'
            ASS.master(want.subtitles)
            ASS.export(want.subtitles)
          when '.ssa'
            SSA.master(want.subtitles)
            SSA.export(want.subtitles)
          when '.srt'
            SRT.master(want.subtitles)
            SRT.export(want.subtitles)
          when '.yt'
            YT.master(want.subtitles)
            YT.export(want.subtitles)
          else
            fail 'Not supported'
          end
      rescue
        @report << "Failure while exporting #{File.extname(want.file)[1..3].upcase} for #{want.file}"
        raise AbortJob
      ensure
        want.subtitles = nil
      end

      if want.encoding
        begin
          data.encode!(want.encoding)
        rescue
          @report << "Failure while transcoding from #{data.encoding} to #{want.encoding} for #{want.file}"
          raise AbortJob
        end
      end

      begin
        IO.write(want.file, data)
      rescue
        @report << "Failure while writing to #{want.file}"
        raise AbortJob
      end
    end

    # Applies a simple timeshift to the subtitle we {Have}.
    # Each subtitle gets shifted forward/backward by the same amount of seconds.
    #
    # @param have [Have] the subtitles we {Have}
    # @param want [Want] the subtitles we {Want}
    # @param reference [Symbol, String] the key of the reference
    def retime_by_single_reference(have, want, reference)
      amount = want.references[reference][:timecode] -
               have.references[reference][:timecode]

      have.subtitles.each do |subtitle|
        subtitle[:start] += amount
        subtitle[:end] += amount
      end
    end

    # Rescales timecodes based on two differing framerates and then applies a
    # simple timeshift to the subtitles we {Have}.
    #
    # @param have [Have] the subtitles we {Have}
    # @param want [Want] the subtitles we {Want}
    # @param reference [Symbol, String] the key of the reference
    def retime_by_framerate_plus_reference(have, want, reference)
      ratio = want.fps.to_f / have.fps.to_f
      have.references[reference][:timecode] *= ratio
      have.subtitles.each do |subtitle|
        subtitle[:start] *= ratio
        subtitle[:end] *= ratio
      end

      amount = want.references[reference][:timecode] -
               have.references[reference][:timecode]

      have.subtitles.each do |subtitle|
        subtitle[:start] += amount
        subtitle[:end] += amount
      end
    end

    # Applies a progressive timeshift on the subtitles we {Have}
    # Two points in time are known for both of which a different shift in time
    # should be applied. Thus a steadily increasing or decreasing forwards or
    # backwards shift will be applied to each subtitle, depending on its
    # place in the time continuum
    #
    # @param have [Have] the subtitles we {Have}
    # @param [Array<Float>] origins the two points in time (given in seconds)
    #   which shall be shifted differently
    # @param [Array<Float>] targets the two amounts of time by which to shift
    #   either of the two points that shall be shifted
    def retime_by_double_reference(have, want, reference_a, reference_b)
      origins = [
        have.references[reference_a][:timecode],
        have.references[reference_b][:timecode]
      ]

      targets = [
        want.references[reference_a][:timecode],
        want.references[reference_b][:timecode]
      ]

      rescale_factor = (targets[1] - targets[0]) / (origins[1] - origins[0])
      rebase_shift = targets[0] - origins[0] * rescale_factor

      have.subtitles.each do |subtitle|
        subtitle[:start] = subtitle[:start] * rescale_factor + rebase_shift
        subtitle[:end] = subtitle[:end] * rescale_factor + rebase_shift
      end
    end

    # Rescales timecodes based on two differing framerates.
    #
    # @param have [Have] the subtitles we {Have}
    # @param want [Want] the subtitles we {Want}
    def retime_by_framerate(have, want)
      ratio = want.fps.to_f / have.fps.to_f
      have.subtitles.each do |subtitle|
        subtitle[:start] *= ratio
        subtitle[:end] *= ratio
      end
    end
  end
end
